File: add-account-information-record.py
# 社員のメールアドレスと名前と登録日付をjson形式のeventという変数で受取り、DynamoDBにレコードとして登録するLambda関数
# このLambda関数は、API GatewayのPOSTメソッドで呼び出されることを想定している
# このLambda関数は、DynamoDBのテーブルにレコードを追加し、アカウント作成のためのLambda関数を呼び出す
import boto3
import json

def lambda_handler(event, context);
    # event変数には、API GatewayのPOSTメソッドで受け取ったjson形式のリクエストが格納されている
    # event変数の中身は以下のようになっている
    # {
    #   "email": "メールアドレス",
    #   "name": "名前",
    #   "registration_date": "登録日付"
    # }
    # この情報をDynamoDBに登録する
    dynamodb = boto3.resource('dynamodb')
    table = dynamodb.Table('account-information')
    # メールアドレスがすでに登録されているかどうかをチェックする
    response = table.get_item(
        Key={
            'email': event['email']
        }
    )
    if 'Item' in response:
        return {
            'statusCode': 400,
            'body': json.dumps('Account information already exists')
        }
    # メールアドレスが登録されていない場合は、DynamoDBにレコードを追加する
    else:
        table.put_item(
            Item={
                'email': event['email'],
                'name': event['name'],
                'registration_date': event['registration_date']
            }
        )
        return {
            'statusCode': 200,
            'body': json.dumps('Account information added successfully')
        }


    

File: script/add-acount-information-record/add-account-information-record.go
package main

import (
    "encoding/json"
    "fmt"
    "net/http"
    "os"

    "github.com/aws/aws-lambda-go/events"
    "github.com/aws/aws-lambda-go/lambda"
    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/dynamodb"
    "github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
)

type Event struct {
    Email           string `json:"email"`
    FirstName       string `json:"first_name"`
    LastName        string `json:"last_name"`
    RegistrationDate string `json:"registration_date"`
}

func handler(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
    var event Event
    err := json.Unmarshal([]byte(request.Body), &event)
    if err != nil {
        return events.APIGatewayProxyResponse{
            StatusCode: http.StatusBadRequest,
            Body:       err.Error(),
        }, nil
    }

    sess := session.Must(session.NewSession())
    svc := dynamodb.New(sess)
    tableName := os.Getenv("TABLE_NAME")

    // Check if the email already exists
    result, err := svc.GetItem(&dynamodb.GetItemInput{
        TableName: aws.String(tableName),
        Key: map[string]*dynamodb.AttributeValue{
            "email": {
                S: aws.String(event.Email),
            },
        },
    })
    if err != nil {
        return events.APIGatewayProxyResponse{
            StatusCode: http.StatusInternalServerError,
            Body:       err.Error(),
        }, nil
    }

    if result.Item != nil {
        return events.APIGatewayProxyResponse{
            StatusCode: http.StatusBadRequest,
            Body:       "Account information already exists",
        }, nil
    }

    // Add the new record
    item, err := dynamodbattribute.MarshalMap(event)
    if err != nil {
        return events.APIGatewayProxyResponse{
            StatusCode: http.StatusInternalServerError,
            Body:       err.Error(),
        }, nil
    }

    _, err = svc.PutItem(&dynamodb.PutItemInput{
        TableName: aws.String(tableName),
        Item:      item,
    })
    if err != nil {
        return events.APIGatewayProxyResponse{
            StatusCode: http.StatusInternalServerError,
            Body:       err.Error(),
        }, nil
    }

    return events.APIGatewayProxyResponse{
        StatusCode: http.StatusOK,
        Body:       "Account information added successfully",
    }, nil
}

func main() {
    lambda.Start(handler)
}